donazioni/models.py

```diff
@@ -0,0 +1,304 @@
+
+    @classmethod
+    def _donatore_esistente(cls, **kwargs):
+        """
+        Ritorna il primo donatore esistente che risponde ai parametri di ricerca
+        None se non esiste alcun record corrispondente
+        :param kwargs:
+        :return: oggetto Donatore o None
+        Stesso comportamento del seguente codice:
+        try:
+            istanza = Donatore.objects.get(**kwargs)
+        except Donatore.DoesNotExist:
+            return
+        except MultipleObjectsReturned:
+            return Donatore.objects.filter(**kwargs).first()
+        else:
+            return istanza
+        """
+        return Donatore.objects.filter(**kwargs).first()
+
+    @classmethod
+    def nuovo_o_esistente(cls, donatore):
+        """
+        Questo metodo contribuisce ad implementare la logica del requisito C-9:
+        'Se sono forniti dati ulteriori dal donatore devono comprendere almeno uno
+        tra i seguenti blocchi di informazioni:
+            Indirizzo e-mail
+            Codice Fiscale (per persone fisiche)
+            Nome, Cognome, Data e Luogo di nascita (per persone fisiche)
+            Ragione sociale e codice fiscale (per persone giuridiche)
+            Nome, Cognome (per persone fisiche)'
+
+        Vedere anche i metodi privati _donatore_esistente e _riconcilia_istanze
+        :param donatore: oggetto Donatore
+        :return: Record esistente o appena creato
+        """
+        istanza = None
+        if donatore.codice_fiscale:
+            if not donatore.ragione_sociale:
+                istanza = cls._donatore_esistente(codice_fiscale=donatore.codice_fiscale)
+            else:
+                istanza = cls._donatore_esistente(codice_fiscale=donatore.codice_fiscale,
+                                                  ragione_sociale=donatore.ragione_sociale)
+            if not istanza and donatore.email:
+                istanza = cls._donatore_esistente(email=donatore.email)
+
+        elif donatore.email:
+            istanza = cls._donatore_esistente(email=donatore.email)
+        elif donatore.nome and donatore.cognome and donatore.data_nascita and donatore.comune_nascita:
+            istanza = cls._donatore_esistente(nome=donatore.nome,
+                                              cognome=donatore.cognome,
+                                              data_nascita=donatore.data_nascita,
+                                              comune_nascita=donatore.comune_nascita)
+        if not istanza:
+            # crea nuovo donatore dai dati derivanti dal modulo ModuloDonatore
+            donatore.save()
+            return donatore
+        # nel caso di donatore esistente, vengono aggiunti nuovi campi se presenti nell'oggetto 'donatore'
+        # e non presenti nell'istanza esistente
+        istanza = cls._riconcilia_istanze(istanza, donatore)
+        return istanza
+
+    @classmethod
+    def _riconcilia_istanze(cls, donatore_esistente, donatore_nuovo):
```

Comment:
Questo metodo manca di un controllo del quale abbiamo discusso telefonicamente. In particolare, e' necessario creare due istanze, piuttosto che riconciliare i record, nel caso in cui tra i due record (esistente, nuovo) vi siano discordanze su campi ausiliari (ovvero, non utilizzati per controllare l'univocita' del record).

Quello che puo' accadere attualmente -con questo codice- e' che due Comitati inseriscano, ad esempio, due indirizzi di residenza diversi, magari perche' il donatore ha cambiato indirizzo tra una donazione a un Comitato e un'altra al secondo. Il comportamento desiderato, avendo analizzato i rischi, e' che i due record relativi al donatore vengano mantenuti separati. Questo evita che il secondo Comitato possa accedere ai dati forniti (come l'indirizzo di residenza passato), attraverso la donazione, al primo Comitato -- dati ai quali, legalmente, potrebbe non avere diritto di accesso.